### PR TITLE
Make inspector polyfill compatible with both null safe and legacy Flutter

### PR DESCRIPTION
**Pull request by jacob314**

### CODE REVIEW

File: packages/devtools_app/assets/scripts/inspector_polyfill_script.dart [hunk 1]
```diff
@@ -94,43 +105,45 @@ String addServiceExtensions() {
...
   }
 
   Future<Map<String, dynamic>> getLayoutExplorerNode(
       Map<String, String> parameters) {
-    final String id = parameters['id'];
-    final int subtreeDepth = int.parse(parameters['subtreeDepth']);
-    final String groupName = parameters['groupName'];
-    Map<String, Object> result = {};
+    final id = parameters['id'];
+    final subtreeDepth = int.parse(parameters['subtreeDepth'] as String);
```
Comment (jacob314) [inline packages/devtools_app/assets/scripts/inspector_polyfill_script.dart:114]: writing `as Foo` is equivalent to writing `!`

File: packages/devtools_app/assets/scripts/inspector_polyfill_script.dart [hunk 2]
```diff
@@ -94,43 +105,45 @@ String addServiceExtensions() {
...
                             includeProperties: true,
                           ),
-                        );
-                final Constraints constraints = renderObject.constraints;
-                if (constraints != null) {
-                  final Map<String, Object> constraintsProperty =
-                      <String, Object>{
+                        ) as Object;
+                try {
+                  final constraints = renderObject.constraints;
```
Comment (jacob314) [inline packages/devtools_app/assets/scripts/inspector_polyfill_script.dart:145]: the constraints property is now non-nullable but could throw an exception if it is accessed at the wrong time..

File: packages/devtools_app/assets/scripts/inspector_polyfill_script.dart [hunk 3]
```diff
@@ -29,6 +29,17 @@
 // additional imports to this file.
 // ignore_for_file: unused_import
 
+// This code needs to execute both in null safe and legacy versions of Dart
+// so it needs to be written in a way that is null safe without depending on any
+// of the new null safety syntax. Te test that this script works with null safe
```
Comment (kenzieschmoll) [inline packages/devtools_app/assets/scripts/inspector_polyfill_script.dart:34]: typo: to

File: packages/devtools_app/assets/scripts/inspector_polyfill_script.dart [hunk 4]
```diff
@@ -94,43 +105,45 @@ String addServiceExtensions() {
...
   Future<Map<String, dynamic>> getLayoutExplorerNode(
       Map<String, String> parameters) {
-    final String id = parameters['id'];
-    final int subtreeDepth = int.parse(parameters['subtreeDepth']);
-    final String groupName = parameters['groupName'];
-    Map<String, Object> result = {};
+    final id = parameters['id'];
+    final subtreeDepth = int.parse(parameters['subtreeDepth'] as String);
+    final groupName = parameters['groupName'];
+    var result = <String, dynamic>{};
```
Comment (kenzieschmoll) [inline packages/devtools_app/assets/scripts/inspector_polyfill_script.dart:116]: why the change from Object to dynamic?

File: packages/devtools_app/assets/scripts/inspector_polyfill_script.dart [hunk 5]
```diff
@@ -94,43 +105,45 @@ String addServiceExtensions() {
...
+          groupName: groupName,
+          summaryTree: true,
+          subtreeDepth: subtreeDepth,
+          service: instance,
+          addAdditionalPropertiesCallback: (node, delegate) {
+            final Map<String, Object> additionalJson = <String, Object>{};
+            final Object value = node.value as Object;
+            if (value is Element) {
+              final renderObject = value.renderObject;
+              if (renderObject != null) {
```
Comment (kenzieschmoll) [inline packages/devtools_app/assets/scripts/inspector_polyfill_script.dart:136]: should this be checking that `renderObject.toDiagnosticsNode != null`?

File: packages/devtools_app/assets/scripts/inspector_polyfill_script.dart [hunk 6]
```diff
@@ -143,88 +156,94 @@ String addServiceExtensions() {
...
-              return additionalJson;
-            }),
-      );
-    }
-    return Future.value(<String, Object>{
+            }
+            return additionalJson;
+          }),
+    ) as Map<String, dynamic>;
+    return Future.value(<String, dynamic>{
```
Comment (kenzieschmoll) [inline packages/devtools_app/assets/scripts/inspector_polyfill_script.dart:181]: again, why the change from Object to dynamic?

File: packages/devtools_app/assets/scripts/inspector_polyfill_script.dart (hunk 3 above)
Comment (jacob314) [reply packages/devtools_app/assets/scripts/inspector_polyfill_script.dart:34]: done

File: packages/devtools_app/assets/scripts/inspector_polyfill_script.dart (hunk 5 above)
Comment (jacob314) [reply packages/devtools_app/assets/scripts/inspector_polyfill_script.dart:136]: Is there a typo in your question? renderObject.toDiagnosticsNode is a Function.
Do you mean checking that the result of toDiagnosticsNode() is not null? It is now enforced as null safe in Flutter and has always been null safe by convention. The previous check was overly paranoid and triggered a lint error with null safety.

File: packages/devtools_app/assets/scripts/inspector_polyfill_script.dart (hunk 6 above)
Comment (jacob314) [reply packages/devtools_app/assets/scripts/inspector_polyfill_script.dart:181]: same answer as above.

File: packages/devtools_app/assets/scripts/inspector_polyfill_script.dart (hunk 4 above)
Comment (jacob314) [reply packages/devtools_app/assets/scripts/inspector_polyfill_script.dart:116]: this has to be dynamic because that is now the type on the Flutter side. Before null safety, Object and dynamic basically meant the same thing so you could use them interchangeably in generics without getting warnings. With null safety they are different because
`Object?` and `Object` are different.

**Comment by jacob314** [comment]
Reverted the pubspec.lock update to avoid a deprecated method issue with the latest vm_service package version.

File: packages/devtools_app/assets/scripts/inspector_polyfill_script.dart [hunk 7]
```diff
@@ -70,8 +81,8 @@ import 'package:flutter/src/widgets/widget_inspector.dart';
 // WidgetInspectorService show up as in scope as far as the analysis server is
 // concerned.
 extension WidgetInspectorServicePrivateMethods on WidgetInspectorService {
-  Map<String, Object> _nodeToJson(
-    DiagnosticsNode node,
+  Map<String, Object /*?*/ > /*?*/ _nodeToJson(
```
Comment (terrylucas) [inline packages/devtools_app/assets/scripts/inspector_polyfill_script.dart:84]: This and below would it be easier to find the null safety code by using a TODO or NNBD type of line comment e.g.,
 // NNBD (jacob):
// Map<String, Object?>?_nodeToJson(
// DiagnosticsNode? node,
// Pre-NNBD(jacob):
Map<String, Object> _nodeToJson(
DiagnosticsNode node,

File: packages/devtools_app/assets/scripts/inspector_polyfill_script.dart [hunk 8]
```diff
@@ -143,88 +156,94 @@ String addServiceExtensions() {
...
                 }
                 if (renderObject is RenderBox) {
                   additionalJson['size'] = <String, Object>{
                     'width': renderObject.size.width.toString(),
                     'height': renderObject.size.height.toString(),
                   };
 
-                  final ParentData parentData = renderObject.parentData;
+                  final ParentData parentData =
+                      renderObject.parentData as ParentData;
```
Comment (terrylucas) [inline packages/devtools_app/assets/scripts/inspector_polyfill_script.dart:169]: `final ParentData` and `as ParentData` don't these do the same?  Only need one?

File: packages/devtools_app/assets/scripts/inspector_polyfill_script.dart [hunk 9]
```diff
@@ -143,88 +156,94 @@ String addServiceExtensions() {
...
+    return Future.value(<String, dynamic>{
       'result': result,
     });
   }
 
   Future<Map<String, dynamic>> setFlexFit(Map<String, String> parameters) {
-    final String id = parameters['id'];
-    final FlexFit flexFit =
-        toEnumEntry<FlexFit>(FlexFit.values, parameters['flexFit']);
+    final String id = parameters['id'] as String;
```
Comment (terrylucas) [inline packages/devtools_app/assets/scripts/inspector_polyfill_script.dart:187]: Same question `final String` and `as String` only one needed?

File: packages/devtools_app/assets/scripts/inspector_polyfill_script.dart [hunk 10]
```diff
@@ -143,88 +156,94 @@ String addServiceExtensions() {
...
     return Future.value(<String, Object>{
       'result': succeed,
     });
   }
 
   Future<Map<String, dynamic>> setFlexFactor(Map<String, String> parameters) {
-    final String id = parameters['id'];
-    final String flexFactor = parameters['flexFactor'];
-    final int factor = flexFactor == 'null' ? null : int.parse(flexFactor);
+    final String id = parameters['id'] as String;
```
Comment (terrylucas) [inline packages/devtools_app/assets/scripts/inspector_polyfill_script.dart:207]: Same question on these two too.

File: packages/devtools_app/assets/scripts/inspector_polyfill_script.dart [hunk 11]
```diff
@@ -143,88 +156,94 @@ String addServiceExtensions() {
...
+        succeed = true;
+      }
     }
     return Future.value({'result': succeed});
   }
 
   Future<Map<String, dynamic>> setFlexProperties(
       Map<String, String> parameters) {
-    final String id = parameters['id'];
+    final String id = parameters['id'] as String;
```
Comment (terrylucas) [inline packages/devtools_app/assets/scripts/inspector_polyfill_script.dart:226]: Same question String and as String?

**Review by terrylucas — APPROVED** [review]

File: packages/devtools_app/assets/scripts/inspector_polyfill_script.dart (hunk 11 above)
Comment (jacob314) [reply packages/devtools_app/assets/scripts/inspector_polyfill_script.dart:226]: parameters['id'] might be null. 
In null safe dart the ideal way to write this is
`final id = parameters['id']!;`
but you can't use `!` and make the code compatible with legacy dart.

File: packages/devtools_app/assets/scripts/inspector_polyfill_script.dart (hunk 10 above)
Comment (jacob314) [reply packages/devtools_app/assets/scripts/inspector_polyfill_script.dart:207]: parameters['id'] might be null. 
In null safe dart the ideal way to write this is
`final id = parameters['id']!;`
but you can't use `!` and make the code compatible with legacy dart.

File: packages/devtools_app/assets/scripts/inspector_polyfill_script.dart (hunk 9 above)
Comment (jacob314) [reply packages/devtools_app/assets/scripts/inspector_polyfill_script.dart:187]: answered lower down.

File: packages/devtools_app/assets/scripts/inspector_polyfill_script.dart (hunk 8 above)
Comment (jacob314) [reply packages/devtools_app/assets/scripts/inspector_polyfill_script.dart:169]: you are right. these types don't need to be on the LHS but they do need to be on the RHS. This code matches flutter style of putting types everywhere so it is easier to merge into package:flutter.

File: packages/devtools_app/assets/scripts/inspector_polyfill_script.dart (hunk 7 above)
Comment (jacob314) [reply packages/devtools_app/assets/scripts/inspector_polyfill_script.dart:84]: I agree that is a bit cleaner. I'll add that refactor in the morning.

**Comment by kenzieschmoll** [comment]
lgtm_thumbs_up
